### DISCUSSION TITLE
Using only 3 buttons - modifying button actions

### DISCUSSION THREAD

**Discussion by Prithul0218** [Spirik/GEM#94] [Q&A] [answered]
Hi,

I am using GEM for a fairly advanced menu with only 3 buttons - Enter, Up (or Left), and Down (or Right). So far I have been able to control everything (some bools and 0-9 integer values) with this. But now I would like to modify some float variables and unfortunately I have no way to go to the left and right digits. 

**Main question:** Is there any way I could use the Enter key to go to the next digit in a float menu item? That would solve my issue. Perhaps when the user presses Enter on the last digit, it gets confirmed. Or even better, if possible, the user can hold the Enter key for 2 seconds to confirm the value. Short presses would just work like the Right key being pressed.

One additional question: Is there any way to set first digit to be edited to the least significant digit (right-most digit)? By default when I hit enter on a float menu item, it starts editing the most significant digit (left-most digit). Changing it could be useful for certain values and menus.

Appreciate any help or advice. Thank you!

**Comment by Prithul0218** [comment]
**Update:** Don't know how I missed[ this discussion](https://github.com/Spirik/GEM/discussions/69). Going to try it first, sorry!

**Update 2:** Tried making GEM._editValueMode public. Enter key now probably triggers Right key too but since Enter key is also being pressed, it comes out of editing the float menu item. So I still need assistance on what I might be missing here.

**Reply by Spirik** [reply, marked as answer]
Hi there!

In order to override default behavior of key presses you have to pass key codes to `menu.registerKeyPress()` method explicitly as described in [Control menu from your sketch](https://github.com/Spirik/GEM/wiki/How-to#control-menu-from-your-sketch) section of wiki. You may want to do that based on the value of `_editValueMode` flag that you've made public (btw, preferred way to do that w/o modification of source code is as described in [Inherit classes](https://github.com/Spirik/GEM/wiki/How-to#inherit-classes) section of wiki). Also be aware of selects, as they behavior should be left as is.

So very roughly it may look something like this (not tested!):
```cpp
// Make protected fields accessible through custom getter methods
class GEMProxy : public GEM_u8g2 {
  public:
    using GEM_u8g2::GEM_u8g2;

    // Make _editValueMode accessible
    bool getEditValueMode() {
      return _editValueMode;
    }

    // Make _editValueType accessible
    byte getEditValueType() {
      return _editValueType ;
    }
	
    // Detect if cursor is at the last digit of editable variable
    bool reachedLastDigit() {
      return (_editValueCursorPosition == _editValueLength - 1) && (_valueString[_editValueCursorPosition] == '\0');
    }
};

// Create menu object as an instance of GEMProxy class
GEMProxy menu(u8g2, GEM_POINTER_ROW, GEM_ITEMS_COUNT_AUTO);

// Later in your loop detect key presses
void loop() {
  // If menu is ready to accept button press...
  if (menu.readyForKey()) {
    // ...detect key press using KeyDetector library (native U8g2 method of key detection can be adapted to this as well)
    myKeyDetector.detect();
    // Pass pressed button to menu based on current state of the menu
    // (pressed button ID is stored in trigger property of KeyDetector object)
    if (myKeyDetector.trigger == GEM_KEY_OK) {
      // If Enter button is pressed while in edit mode...
      if (menu.getEditValueMode() && (menu.getEditValueType() != GEM_VAL_SELECT)) {
        // Menu currently is in edit mode, treat button either as GEM_KEY_RIGHT or GEM_KEY_OK (depending on cursor position)
        menu.registerKeyPress(menu.reachedLastDigit ? GEM_KEY_OK : GEM_KEY_RIGHT);
      } else {
        // ...else as GEM_KEY_OK
        menu.registerKeyPress(GEM_KEY_OK);
      }
    } else {
      // ...else detect other buttons as usual
      menu.registerKeyPress(myKeyDetector.trigger);
    }
  }
}
```

Another possible solution is to track duration and/or combination of button presses. Similar to how it is done in [Rotary encoder examples](https://github.com/Spirik/GEM/wiki/Test-bench-with-rotary-encoder-for-U8g2-version) (rotary encoder essentially has three buttons: rotation left, rotation right and push button). That way it won't be necessary to override GEM classes, but you would have to deal with custom `millis()` based detection of key presses (as seen [here](https://github.com/Spirik/GEM/blob/master/examples/U8g2/Example-05_Encoder/Example-05_Encoder.ino#L137)).

**Reply by Prithul0218** [reply]
Thank you so much for the detailed answer Spirik! I had to make some minor changes to your code like changing `reachedLastDigit()` and `getEditValueType()` function return types to `uint8_t` and it just worked! 
Being only an intermediate coder in C++, I did not know it was possible to "proxy" a class like that, pretty cool feature and I'm glad I learnt something new!

One quality of life improvement feature I would like to add though is to be able to start editing floats from the last digit (LSB). It is useful for my project to only make small changes most often. I made some progress on it by creating a function:
```
bool setEditValueCursorPosition(int position) {
        _editValueCursorPosition = position;
        _editValueVirtualCursorPosition = position;
        return true;
    }
```
and setting it to position 4 (my required float precision for now). However, since `initEditValueCursor()` function under `GEM` class sets `_editValueCursorPosition` to 0, the value starts editing from cursor position 0 and jumps to 4 only after I click the `LEFT` key once. Is there any way to work around it without modifying the library? I thought of simulating a `LEFT` key press to force it to jump to the last cursor position but that did not work for me.

I can't think of a way to modify the `initEditValueCursor()` function from my GEMProxy class. Can I?

Please let me know if you have any advice. Thank you again!

**Reply by Spirik** [reply]
Initial cursor position is set internally in [`GEM_u8g2::initEditValueCursor()`](https://github.com/Spirik/GEM/blob/master/src/GEM_u8g2.cpp#L700) method. I believe, there might be a couple of ways to go about it.

First, it is possible to override this method and set cursor positions to a custom value. To override internal methods you would have to enable [Advanced mode](https://github.com/Spirik/GEM?tab=readme-ov-file#advanced-mode) first and then check out [Override methods in Advanced Mode](https://github.com/Spirik/GEM/wiki/How-to#override-methods-in-advanced-mode) section in wiki. This solution is pretty similar to what we did earlier when writing custom methods to access internal properties (we will rewrite existing one here instead).

The second possible solution doesn't require to enable Advanced Mode in config. As you may notice by looking in a source code of `initEditValueCursor()`, it calls `drawMenu()` method at the very end. In GEM it is possible to specify custom callback function that will be called each time `drawMenu()` is invoked (see [Add `drawMenu()` callback](https://github.com/Spirik/GEM/wiki/How-to#add-drawmenu-callback) section in wiki). However, `drawMenu()` will also be called each time some updates are happening on the screen (when changing cursor position or digit value, etc.). So you would have to come up with some clever solution that will allow you to set desired values only once when edition of variable is started. That might not be trivial (e.g. adding global boolean flag that will be set at the end of your callback and reset when edition of variable is finished in its own [callback](https://github.com/Spirik/GEM/wiki/How-to#add-menu-item-callback-basic) function).

And lastly, probably there still can be a way to achieve that w/o the need for all of the above modifications, by further adjusting your control conditionals in a `loop()` function (e.g. calling `menu.registerKeyPress(GEM_KEY_RIGHT)` several times in a row to move cursor to a desired position).

And I am not sure I understand correctly how user will be able to navigate to the previous digits in that case (if it will be necessary to update more significant digits of a number), because pressing Enter button moves cursor only in one direction (to the right) in the proposed 3-button scheme?

PS: Btw, thank you for corrections, I will update code in my previous answer!

**Comment by lfoxdev** [comment]
I have a 4-button module (which I use as Up, Down, Ok & Cancel) and the same problem, so here is the wrapper i wrote. Numeric values and strings can be edited.
Pressing Ok moves the cursor to the right or saves the value if the cursor is on the last character; pressing Cancel moves the cursor to the left and saves the value if the cursor is on the first character (you can replace the second saveEditValue() with cancelEditValue() to cancel changes).

```CPP
class GEM_u8g2_4keys : public GEM_u8g2 {
  public:
    using GEM_u8g2::GEM_u8g2;

    GEM_u8g2& registerKeyPress(byte keyCode) {
      if (context.loop == nullptr && _editValueMode && _editValueType != GEM_VAL_SELECT && _editValueType != GEM_VAL_SPINNER) {
        if (keyCode == GEM_KEY_OK) {
          if ((_editValueVirtualCursorPosition == _editValueLength - 1) || (_valueString[_editValueVirtualCursorPosition] == '\0'))
            saveEditValue();
          else
            nextEditValueCursorPosition();
        }
        else if (keyCode == GEM_KEY_CANCEL) {
          if (_editValueVirtualCursorPosition == 0)
            saveEditValue();
          else
            prevEditValueCursorPosition();          
        }
        else
          GEM_u8g2::registerKeyPress(keyCode);
      } 
      else
        GEM_u8g2::registerKeyPress(keyCode);
      return *this;
    }

};

GEM_u8g2_4keys menu(u8g2, GEM_POINTER_ROW, GEM_ITEMS_COUNT_AUTO);
```
